packages/headless-components/components/src/react/address.tsx

```diff
@@ -0,0 +1,266 @@
+// ==========================================
+
+enum TestIds {
+  // Root container
+  addressRoot = 'address-root',
+  // Formatted address
+  addressFormatted = 'address-formatted',
+  // Address lines array
+  addressLines = 'address-lines',
+}
+
+// ==========================================
+// Type Definitions
+// ==========================================
+
+/**
+ * Address data structure compatible with @fragaria/address-formatter
+ * Maps common address fields to the formatter's expected structure
+ */
+type AddressData = Parameters<typeof addressFormatter.format>[0];
+
+/**
+ * Formatting options for the address formatter
+ */
+type AddressFormattingOptions = Omit<
+  Parameters<typeof addressFormatter.format>[1],
+  'output'
+> & { output?: 'string' | 'array' };
+
+// ==========================================
+// Context
+// ==========================================
+
+interface AddressContextValue {
+  addressData: AddressData;
+}
+
+const AddressContext = React.createContext<AddressContextValue | null>(null);
+
+function useAddressContext(): AddressContextValue {
+  const context = React.useContext(AddressContext);
+  if (!context) {
+    throw new Error(
+      'useAddressContext must be used within an Address.Root component',
+    );
+  }
+  return context;
+}
+
+// ==========================================
+// Utility Functions
+// ==========================================
+
+/**
+ * Format address using @fragaria/address-formatter
+ */
+function formatAddressString(
+  addressData: AddressData,
+  options: AddressFormattingOptions = {},
+): string {
+  return addressFormatter.format(addressData, {
+    ...options,
+    output: 'string',
+  });
+}
+
+/**
+ * Get address as array of lines using @fragaria/address-formatter
+ */
+function formatAddressLines(
+  addressData: AddressData,
+  options: AddressFormattingOptions = {},
+): string[] {
+  return addressFormatter.format(addressData, {
+    ...options,
+    output: 'array',
+  });
+}
+
+// ==========================================
+// Components
+// ==========================================
+
+// Address.Root Component Props
+interface AddressRootProps {
+  address: AddressData;
+  children: React.ReactNode;
+}
+
+// Address.Root Component
+export const Root = React.forwardRef<HTMLElement, AddressRootProps>(
+  (props, ref) => {
+    const { address, children } = props;
+
+    const contextValue: AddressContextValue = {
+      addressData: address,
+    };
+
+    const attributes = {
+      'data-testid': TestIds.addressRoot,
+      'data-country-code': address.countryCode,
+    };
+
+    return (
+      <AddressContext.Provider value={contextValue}>
+        <AsChildSlot {...attributes} ref={ref}>
+          {children}
+        </AsChildSlot>
+      </AddressContext.Provider>
+    );
+  },
+);
+
+Root.displayName = 'Address.Root';
+
+// Address.Formatted Component Props
+interface AddressFormattedProps {
+  children?: AsChildChildren<{
+    formattedAddress: string;
```

Comment:
when using asChild with render props (function) also return the original address object, this way Event.Address can render formatted address by default but also allow the user to use render props with the address object instead (like parts in intl)
I would even call it `parts`